src/XR/features/WebXRNearInteraction.ts

```diff
@@ -0,0 +1,590 @@
+            }
+        }
+        // NOT SURE: if i should return null for gaze and screen modes
+        switch (xrController.inputSource.targetRayMode) {
+            case "tracked-pointer":
+                return this._attachTrackedPointerRayMode(xrController);
+            case "gaze":
+                return null;
+            case "screen":
+                return null;
+        }
+    };
+
+    private _controllers: {
+        [controllerUniqueId: string]: {
+            xrController?: WebXRInputSource;
+            squeezeComponent?: WebXRControllerComponent;
+            selectionComponent?: WebXRControllerComponent;
+            onButtonChangedObserver?: Nullable<Observer<WebXRControllerComponent>>;
+            onSqueezeButtonChangedObserver?: Nullable<Observer<WebXRControllerComponent>>;
+            onFrameObserver?: Nullable<Observer<XRFrame>>;
+            meshUnderPointer: Nullable<AbstractMesh>;
+            pick: Nullable<PickingInfo>;
+            id: number;
+            pickIndexMeshTip: Nullable<AbstractMesh>;
+            hoverIndexMeshTip: Nullable<AbstractMesh>;
+            grabRay: Ray;
+            nearInteraction: boolean;
+            nearGrab: boolean;
+            nearGrabInProcess: boolean;
+            // event support
+            eventListeners?: { [event in XREventType]?: (event: XRInputSourceEvent) => void };
+        };
+    } = {};
+    private _scene: Scene;
+
+    private _attachedController: string;
+
+    private _farInteractionFeature: Nullable<IWebXRFeature>;
+
+    /**
+     * The module's name
+     */
+    public static readonly Name = WebXRFeatureName.NEAR_INTERACTION;
+    /**
+     * The (Babylon) version of this module.
+     * This is an integer representing the implementation version.
+     * This number does not correspond to the WebXR specs version
+     */
+    public static readonly Version = 1;
+
+    /**
+     * constructs a new background remover module
+     * @param _xrSessionManager the session manager for this module
+     * @param _options read-only options to be used in this module
+     */
+    constructor(_xrSessionManager: WebXRSessionManager, private readonly _options: IWebXRNearInteractionOptions) {
+        super(_xrSessionManager);
+        this._scene = this._xrSessionManager.scene;
+    }
+
+    /**
+     * attach this feature
+     * Will usually be called by the features manager
+     *
+     * @returns true if successful.
+     */
+    public attach(): boolean {
+        if (!super.attach()) {
+            return false;
+        }
+
+        this._options.xrInput.controllers.forEach(this._attachController);
+        this._addNewAttachObserver(this._options.xrInput.onControllerAddedObservable, this._attachController);
+        this._addNewAttachObserver(this._options.xrInput.onControllerRemovedObservable, (controller) => {
+            // REMOVE the controller
+            this._detachController(controller.uniqueId);
+        });

```

Comment:
Why not just have _detachController take in a controller instead of an id?

Reply:
please refer to detach controller function. It is because the id is required to verify if it is attachedController or not (at the bottom of the function).

Reply:
I really dont find big difference passing a controller vs id in this context